### DISCUSSION TITLE
Ludwig for Fantacalcio

### DISCUSSION THREAD

**Discussion by emmecicubo** [ludwig-ai/ludwig#1232] [Q&A] [answered]
Hello,
first of all I would like to thank you for publishing such an amazing tool. I think that I may one of the main targets for it. In fact I am very interested in the topic (next year I am going to study data science), but my knowledge is pretty poor. This is why I would like to ask a couple of questions regarding how Ludwig works.

I would like to employ Ludwig for Fantacalcio (Italian fantasy football, or soccer, depending on where you leave ;)  ).

My idea would be to structure my dataset in this fashion:

Player1, height, age, weight, team, goals, 2020, etc.
Player1, height, age, weight, team, goals, 2019, etc
Player2, height, age, weight, team, goals, 2020, etc

with the dependent variable being the yearly average fantasy score (and maybe the number of match played).

So, here are the questions:

1. Will Ludwig recognize that Player1 2020 and Player1 2019 are the same person in two different moments (and in case not, would that be an issue)?
2. Some players may perform better if they are paired with particulars coaches or players. For example an attacker could score more goals with a coach that employ an offensive game. Let’s suppose that Player1 scored many goals in 2020 because Player2 is very good at passing the ball. Now let’s suppose Player2 goes to another team. I would expect Player1 to perform worse in 2021. Does Ludwig understands this (considering that there is a column that stores the team), or should I make like 30 columns that include the names of all of the teammates?
3. Would I be able to make predictions at the beginning of the season using this historical data, considering that many lines (such as number of played games, goals etc) for the 2021/2022 season would obviously be empty? 

Thank you very much in advance for the answers

**Comment by w4nderlust** [comment, marked as answer]
Hi @emmecicubo !

This is a fun topic. I'm not an expert of fantasy football, but generally speaking I believe that this may work. Pretty sure that most football clubs have systems like what you are describing for evaluating prospect players, although they likely integrate a lot of additional information, but that suggests that you could get something out of it, or at the very least you can brag with your fellow Fantacalcio players about it :)

To answer your questions:

1. Depending on how you represent them, it can do it. In particular if you treatthe player name as their identifier, using the `category` datat type, then yes, the same player name will be recognized across the different years.
2. the deep learning model that ludwig will train can figure out correlations like the ones you are describing, although roughly speaking, the more complex the correlation, the more data you will need to the model to understand it. If you were using data from every single match and every single training session for instance, i'd say it would be quite likely the model will pick something like that up. My humble opinon is that you may want to start training a model without all the other players, and then try to add them (making sure you use the same weights for all the columns containing players) to see if it improves performance, so be empirical about it. One more thing to consider is that maybe the team manager could also be a very influential feature (some players perform better under some managers).
3. If you treat hte year as a numerical feature it's likely the model will pick trajectories, like players that have improved performances in the last two years are more likely to keep improving than players that have been declining, and things like that. Ideally a machine learning model should be capable to extrapolate something from the data and apply it tounseen scenarios, although it isn't always the case :)

Hpefully this helps you.
One more thing I'd like to add is that you may want to enrich the dataset with additional features like:

- team last year, goals last year, average score last year, std score last year
- team two years ago, goals two years ago, average score two years go, std score two years ago

etc. for the last few years.

That may help the model understanding the performance trajectory of the player. And then the target would not be "performance in 2022" but would be "performance current year" so you can go back in time and predict their performance 3 years ago from data from 4 5 6 years ago. It's the same data you are describing but rolled up differently.

Let me know if this works! It would be a really fun example to showcase on the Ludwig website in case.

**Reply by emmecicubo** [reply]
Thank you very much for the answer! It has been very helpful!
There is only one thing that I didn't understand very well.
In the "improvements" part of your reply, are you suggesting to use a single row for each player?

- Player1, team last year, goals last year, average score last year, std score last year, team two years ago, goals two years ago, average score two years go, std score two years ago, etc
- Player2, team last year, goals last year, average score last year, std score last year, team two years ago, goals two years ago, average score two years go, std score two years ago, etc

or just to use the column "2 years ago" instead of "2019" and maintaining multiple rows for every player?
Wouldn't that be redundant in a sense? In this way I would state multiple times that in 2018 player1 had a fantasy score of x (in the 2018 row for the current year, in 2019 for the last year and in 2020 for 2 years ago).

As soon as I gather all of the data I will try the solutions and keep you posted!
Thank you very much!

**Reply by w4nderlust** [reply]
You are welcome!

To answer your question, in the way I was thinking about it, you would still have multiple entries for the mae player, but for different widnows.
For instance, let's assume you have data for Locatelli for 2015 to 2021, and your window is 3 years, you would have one row with data from 2015 (year-2) 2016 (year-1) and 2017 (current year) predicting 2018 (next year), one with data from 2016 2017 and 2018 predicting 2019 and so on. Hopefully this is clearer now.

**Reply by emmecicubo** [reply]
It is! Thank you very much!
